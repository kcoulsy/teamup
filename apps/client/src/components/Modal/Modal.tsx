import React from 'react';

import Portal from '../Portal/Portal';

export interface ModalAction {
  buttonLabel: string;
  onClick: Function;
  klasses?: string;
}

interface ModalProps {
  isActive: boolean;
  headerTitle?: string | JSX.Element;
  actions?: ModalAction[];
}

const Modal: React.FunctionComponent<ModalProps> = ({
  isActive = false,
  children,
  headerTitle,
  actions = [],
}) => {
  return (
    <Portal>
      <div
        className={`ui dimmer modals page visible ${isActive ? ' active' : ''}`}
        style={{ display: 'flex !important' }}>
        <div
          className='ui standard test modal visible active front'
          style={{ display: 'flex !important' }}>
          {headerTitle ? <div className='header'>{headerTitle}</div> : null}
          <div className='content'>{children}</div>
          <div className='actions'></div>
          {actions.length ? (
            <div className='actions'>
              {actions.map((actionObj, i) => {
                let klasses = 'ui button';
                const key = actionObj.buttonLabel + '_' + i;

                if (actionObj.klasses) klasses += ` ${actionObj.klasses}`;

                const handleClick = (
                  event: React.MouseEvent<HTMLDivElement>
                ) => {
                  event.preventDefault();
                  actionObj.onClick();
                };

                return (
                  <div key={key} className={klasses} onClick={handleClick}>
                    {actionObj.buttonLabel}
                  </div>
                );
              })}
            </div>
          ) : null}
        </div>
      </div>
    </Portal>
  );
};

export default Modal;
