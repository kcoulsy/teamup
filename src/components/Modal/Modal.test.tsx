import React from 'react';
import { shallow, mount } from 'enzyme';
import { describe, expect, test, vi } from 'vitest';
import { shallowToJson } from 'enzyme-to-json';
import Modal from './Modal';

describe('modal component', () => {
  test('should render modal correctly', () => {
    const wrapper = shallow(<Modal isActive={false} />);

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  test('should be active if isActive is true', () => {
    const wrapper = shallow(<Modal isActive />);

    expect(shallowToJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.page.active').length).toBe(1);

    wrapper.setProps({ isActive: false });
    expect(wrapper.find('.page.active').length).toBe(0);
  });

  test('should render children', () => {
    const wrapper = shallow(
      <Modal isActive>
        <p>JSX Child</p>
      </Modal>
    );

    expect(wrapper.contains(<p>JSX Child</p>)).toBe(true);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  test('should render headerTitle', () => {
    const title = 'test title';
    const wrapper = shallow(<Modal isActive headerTitle={title} />);

    expect(wrapper.contains(title)).toBe(true);
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    const titleWithJSX = <h4>Test Title</h4>;
    const wrapperWithJSX = shallow(
      <Modal isActive headerTitle={titleWithJSX} />
    );

    expect(wrapperWithJSX.contains(titleWithJSX)).toBe(true);
    expect(shallowToJson(wrapperWithJSX)).toMatchSnapshot();
  });

  test('should render buttons when actions are passed', () => {
    let count = 0;
    const incLabel = 'Inc';
    const decLabel = 'Dec';
    const actions = [
      {
        buttonLabel: incLabel,
        onClick: () => {
          count++;
        },
        klasses: 'testbutton',
      },
      {
        buttonLabel: decLabel,
        onClick: () => {
          count--;
        },
      },
    ];

    const wrapper = shallow(<Modal isActive actions={actions} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    const buttonWrapper = wrapper.find('.testbutton');

    expect(buttonWrapper.length).toBe(1);
    expect(buttonWrapper.contains(incLabel)).toBe(true);
    expect(count).toBe(0);
    buttonWrapper.simulate('click', { preventDefault: vi.fn() });
    expect(count).toBe(1);
  });
});
