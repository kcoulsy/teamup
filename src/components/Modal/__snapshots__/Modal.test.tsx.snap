// Vitest Snapshot v1

exports[`modal component > should be active if isActive is true 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`modal component > should render buttons when actions are passed 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
      <div
        className="actions"
      >
        <div
          className="ui button testbutton"
          key="Inc_0"
          onClick={[Function]}
        >
          Inc
        </div>
        <div
          className="ui button"
          key="Dec_1"
          onClick={[Function]}
        >
          Dec
        </div>
      </div>
    </div>
  </div>
</Portal>
`;

exports[`modal component > should render children 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      >
        <p>
          JSX Child
        </p>
      </div>
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`modal component > should render headerTitle 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="header"
      >
        test title
      </div>
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`modal component > should render headerTitle 2`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="header"
      >
        <h4>
          Test Title
        </h4>
      </div>
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`modal component > should render modal correctly 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible "
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component > should be active if isActive is true 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component > should render buttons when actions are passed 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
      <div
        className="actions"
      >
        <div
          className="ui button testbutton"
          key="Inc_0"
          onClick={[Function]}
        >
          Inc
        </div>
        <div
          className="ui button"
          key="Dec_1"
          onClick={[Function]}
        >
          Dec
        </div>
      </div>
    </div>
  </div>
</Portal>
`;

exports[`render modal component > should render children 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      >
        <p>
          JSX Child
        </p>
      </div>
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component > should render correctly 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible "
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component > should render headerTitle 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="header"
      >
        test title
      </div>
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component > should render headerTitle 2`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="header"
      >
        <h4>
          Test Title
        </h4>
      </div>
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component > should render modal correctly 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible "
    style={
      {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component should be active if isActive is true 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      Object {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        Object {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component should render buttons when actions are passed 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      Object {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        Object {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
      <div
        className="actions"
      >
        <div
          className="ui button testbutton"
          key="Inc_0"
          onClick={[Function]}
        >
          Inc
        </div>
        <div
          className="ui button"
          key="Dec_1"
          onClick={[Function]}
        >
          Dec
        </div>
      </div>
    </div>
  </div>
</Portal>
`;

exports[`render modal component should render children 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      Object {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        Object {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      >
        <p>
          JSX Child
        </p>
      </div>
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component should render correctly 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible "
    style={
      Object {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        Object {
          "display": "flex !important",
        }
      }
    >
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component should render headerTitle 1`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      Object {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        Object {
          "display": "flex !important",
        }
      }
    >
      <div
        className="header"
      >
        test title
      </div>
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;

exports[`render modal component should render headerTitle 2`] = `
<Portal>
  <div
    className="ui dimmer modals page visible  active"
    style={
      Object {
        "display": "flex !important",
      }
    }
  >
    <div
      className="ui standard test modal visible active front"
      style={
        Object {
          "display": "flex !important",
        }
      }
    >
      <div
        className="header"
      >
        <h4>
          Test Title
        </h4>
      </div>
      <div
        className="content"
      />
      <div
        className="actions"
      />
    </div>
  </div>
</Portal>
`;
